### ISSUE TITLE
Test failures possibly due to use of legacy TLS with PyPI

### ISSUE THREAD

**Issue by nathanielmanistaatgoogle** [grpc/grpc#14815]
[Two recent test executions](https://github.com/grpc/grpc/pull/14652) ([1](https://source.cloud.google.com/results/invocations/4b7246c5-77f1-4f25-a69a-314a1bf649f0/targets/grpc%2Fcore%2Fpull_request%2Fmacos%2Fgrpc_basictests_opt/log), [2](https://source.cloud.google.com/results/invocations/83950c02-1e8e-4daa-8b8d-628eb0c1e7f8/targets/grpc%2Fcore%2Fpull_request%2Fmacos%2Fgrpc_basictests_dbg/log)) failed and the message "Brownout of Legacy TLS" appeared several times in their logs. [This is somewhat expected, especially if our tests ran in the first ten minutes of an hour.](https://mail.python.org/pipermail/distutils-sig/2018-March/032068.html)

I suspect we need to upgrade... something, but I'm not yet sure what. Something in the continuous integration infrastructure? [`pip`](https://github.com/grpc/grpc/issues/14814)?

**Comment by matt-kwong** [comment]
I'm not sure if this is something that can be fixed with a macOS version upgrade, but you could test this yourself if you had a macOS computer with the latest OS by running the command in my earlier comment with the system's preinstalled Python. 

I think the issue is that there are two versions of Python 2.7 available on the macOS workers: one installed by brew and the other preinstalled by the system. The latter uses the older TLS version and is the one used for running Python tests. We do a few `pip install x` commands as prebuild steps and have had no issues with those. The binaries we use for the prebuild steps are:
```
++ which python2
/usr/local/bin/python2
++ which pip
/usr/local/bin/pip
```

**Comment by dgquintas** [comment]
This continues to happen. Is there a plan to address it?

**Comment by kpayson64** [comment]
@matt-kwong @dgquintas 
This is something that needs to be forwarded to the kokoro team.

We need the kokoro machines to have a Python 3 and Python 2 version that have TLS 1.2 (as the default version).

We may need to upgrade our run_tests.py matrix to use a different version of Python 3 (likely 3.5 or 3.6), but we can make that change on the advisement of the kokoro team.

**Comment by jtattermusch** [comment]
95% of our mac builds on master are broken because of this, what's the ETA for fixing?

**Comment by matt-kwong** [comment]
Tried to do a force-link of the Brew installed Pythons over the system installed in #14933, but that didn't work. I suspect it's an issue with virtualenv creating a new Python executable. More context in https://github.com/grpc/grpc/pull/14933#issuecomment-378765390.

**Comment by mehrdada** [comment]
How about instead of setting the `PATH` environment explicitly so that brew python precedes the system python instead of symlinks?

**Comment by matt-kwong** [comment]
I think it already does that, because when you run `whereis python`, it points to `/usr/local/bin/`, which is the Brew installed one, and not the system installed one which is in `/usr/bin/`.

**Comment by matt-kwong** [comment]
I've tried a few workarounds that haven't been successful, including:
- hardcoding the python path in `build_python.sh` when creating venv
- force-linking the Brew installed Python over the system installed Python
- Using the system pip instead of the venv pip (this one doesn't work for other reasons)

I'm going to focus on the upgrade to High Sierra, which I think should fix this issue.

**Comment by mehrdada** [comment]
I am bumping priority on this... It is actually making release builds fail.

**Comment by nathanielmanistaatgoogle** [comment]
Further external discussion: ["Impending silent breakage of pip / macOS likely to cause severe confusion"](https://groups.google.com/d/topic/pypa-dev/Oz6SGA7gefo).

**Comment by matt-kwong** [comment]
I might have a fix. I tried upgrading pip in our macOS setup script, but that originally didn't work. I didn't realize in [`build_python.sh`](https://github.com/grpc/grpc/blob/master/tools/run_tests/helper_scripts/build_python.sh#L161) that pip was being downgraded to 9.0.1 after I upgraded pip. I'm going to test changing that pin to 9.0.3 and see if that fixes the issue.

**Comment by matt-kwong** [comment]
```
+ /Volumes/BuildData/tmpfs/src/github/grpc/workspace_python_macos_opt_native/py34_native/bin/python3.4 -m pip install --upgrade pip==9.0.3
Collecting pip==9.0.3
  Could not find a version that satisfies the requirement pip==9.0.3 (from versions: )
No matching distribution found for pip==9.0.3
You are using pip version 9.0.1, however version 9.0.3 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
```
... well okay then.

**Comment by matt-kwong** [comment]
https://source.cloud.google.com/results/invocations/628313ee-732e-45e9-9d8c-4878ee92f89d/targets - Python tests are passing if I unpin the pip version. Reusing #14956 to unpin the pip version in `build_python.sh`

**Comment by mehrdada** [comment]
@matt-kwong Can we pin the min version at least instead of letting it loose? Will it keep working?

**Comment by matt-kwong** [comment]
@mehrdada I tried pinning to 9.0.3 but that failed. I'll try 9.0.2. Is there a reason why we prefer to pin pip? It doesn't get get updated often (only 2 this year and none last year), and I'm not sure if it's the kind of dependency that breaks our tests often.

**Comment by mehrdada** [comment]
I think there used to be an issue on systems with very old pip--that pin was added long before my time. I don't have much historical context, but @nathanielmanistaatgoogle and @kpayson64 may have more to add. I don't have a strong opinion either way.

**Comment by nathanielmanistaatgoogle** [comment]
@matt-kwong: dependency-pinning is a practice I brought with me from my previous project after too many experiences seeing my own build break because someone elsewhere on the Internet did something bad. With pinning the decision to break my build remains with someone elsewhere on the Internet, but the _timing_ of breaking my build falls to me and my decision to move my pin, so it's not a perfect system but it is an improvement. :-)

**Comment by matt-kwong** [comment]
Okay, pinning to 9.0.2 works, but not 9.0.3 (I'm guessing this is a pip issue). It's fine though because the fix we need is in 9.0.2. I'll merge the PR that unpins and open another PR to pin to 9.0.2; I want to get the release blocker in sooner rather than later.

**Comment by matt-kwong** [comment]
Merged #14956

**Comment by matt-kwong** [comment]
Reopening because seeing 
```
+ /Volumes/BuildData/tmpfs/src/github/grpc/workspace_python_macos_opt_native/py34_native/bin/python3.4 -m pip install --upgrade pip==9.0.2
Collecting pip==9.0.2
  Could not fetch URL https://pypi.python.org/simple/pip/: There was a problem confirming the ssl certificate: [SSL: TLSV1_ALERT_PROTOCOL_VERSION] tlsv1 alert protocol version (_ssl.c:600) - skipping
  Could not find a version that satisfies the requirement pip==9.0.2 (from versions: )
No matching distribution found for pip==9.0.2
You are using pip version 9.0.1, however version 9.0.3 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
2018-04-12 09:07:51,382 FAILED: /Volumes/BuildData/tmpfs/src/github/grpc/workspace_python_macos_opt_native/tools/run_tests/helper_scripts/build_python.sh [ret=1, pid=80593, time=3.4sec]
2018-04-12 09:07:51,388 FAILED: Some tests failed
2018-04-12 09:07:51,420 FAILED: run_tests_python_macos_opt_native [ret=1, pid=69434, time=429.4sec]
```
which is the same issue but failing one step earlier.
